crawl-ref/source/util/job-gen.py

```diff
@@ -0,0 +1,342 @@
+#!/usr/bin/env python
+
+# --- WORK IN PROGRESS --- #
+# Aim: to generate job-type.h and job-data.h in the same way as species,
+# from yaml files
+
+# NOTE Wanderers have empty skills and equipment sets
+
+from __future__ import print_function
+
+import argparse
+import os
+import sys
+import traceback
+import collections
+import re
+
+import yaml  # pip install pyyaml
+
+def quote_or_nullptr(key, d):
+    if key in d:
+        return quote(d[key])
+    else:
+        return 'nullptr'
+
+class Job(collections.MutableMapping):
+    """Parser for YAML definition files.
+
+    If any YAML content is invalid, the relevant parser function below should
+    raise ValueError.
+    """
+
+    # TODO: unify with processing in from_yaml
+    YAML_MAIN_FIELDS = {'TAG_MAJOR_VERSION', # 'category', 'category_priority' 
+            'recommended_species', 'enum', 'name', 'short_name', 'str', 'int',
+            'dex', 'weapon_choice', 'create_enum', 'skills', 'equipment'}
+
+    def __init__(self, yaml_dict):
+        self.backing_dict = dict()
+        self.from_yaml(yaml_dict)
+
+    def __getitem__(self, key):
+        return self.backing_dict[key]
+
+    def __setitem__(self, key, val):
+        self.backing_dict[key] = val
+
+    def __delitem__(self, key):
+        del self.backing_dict[key]
+
+    def __iter__(self):
+        return iter(self.backing_dict)
+
+    def __len__(self):
+        return len(self.store)
+
+    def print_unknown_warnings(self, s):
+        for key in s:
+            if key not in self.YAML_MAIN_FIELDS:
+                print("species_gen.py warning: Unknown field '%s' in species %s"
+                                    % (key, self['enum']), file=sys.stderr)
+
+    def from_yaml(self, s):
+        # Pre-validation
+        if s.get('TAG_MAJOR_VERSION', None) is not None:
+            if not isinstance(s['TAG_MAJOR_VERSION'], int):
+                raise ValueError('TAG_MAJOR_VERSION must be an integer')
+    #    self.starting_species = s.get('difficulty') != False
+    #    has_recommended_species = bool(s.get('recommended_species'))
+    #    if self.starting_species != has_recommended_species:
+    #        raise ValueError('recommended_jobs must not be empty (or'
+    #                                            ' difficulty must be False)')
+
+        # Set attributes
+        self['enum'] = validate_string(s['enum'], 'enum', 'JOB_[A-Z_]+$')
+        self['name'] = validate_string(s['name'], 'name', '..+')
+        self['short_name'] = s.get('short_name', s['name'][:2])
+        # TODO: currently done by skill enum values; change to be strings?
+        self['skills'] = skills(s.get('skills', {}))
+        self['str'] = validate_int_range(s['str'], 'str', -15, 15)
```

Comment:
This might be overreach, but maybe str+int+dex should be validated to equal 0 or 12.

Reply:
I wondered about this. From the point of view of adding new backgrounds, where fork developers may choose to stray from this rule though, I thought it best to leave it.